python/ray/_private/node.py

```diff
@@ -477,65 +517,62 @@ def sigterm_handler(signum, frame):
 
         ray._private.utils.set_sigterm_handler(sigterm_handler)
 
-    def _init_temp(self):
+    def _init_temp(self, node_to_connect_info):
         # Create a dictionary to store temp file index.
         self._incremental_dict = collections.defaultdict(lambda: 0)
 
-        if self.head:
-            self._ray_params.update_if_absent(
-                temp_dir=ray._common.utils.get_ray_temp_dir()
-            )
-            self._temp_dir = self._ray_params.temp_dir
-        else:
-            if self._ray_params.temp_dir is None:
-                assert not self._default_worker
-                temp_dir = ray._private.utils.internal_kv_get_with_retry(
-                    self.get_gcs_client(),
-                    "temp_dir",
-                    ray_constants.KV_NAMESPACE_SESSION,
-                    num_retries=ray_constants.NUM_REDIS_GET_RETRIES,
-                )
-                self._temp_dir = ray._common.utils.decode(temp_dir)
+        self.temp_dir = self._ray_params.temp_dir
+        if self.temp_dir is None:
+            if node_to_connect_info is not None:
+                self.temp_dir = node_to_connect_info.temp_dir
             else:
-                self._temp_dir = self._ray_params.temp_dir
-
-        try_to_create_directory(self._temp_dir)
-
-        if self.head:
-            self._session_dir = os.path.join(self._temp_dir, self._session_name)
+                if self.head:
+                    # fallback to head node's default
+                    self.temp_dir = ray._common.utils.get_default_ray_temp_dir()
+                else:
+                    # fallback to head node's temp dir if no node info is found for the given node ip address
+                    assert not self._default_worker
+                    self.temp_dir = self._head_temp_dir
         else:
-            if self._temp_dir is None or self._session_name is None:
-                assert not self._default_worker
-                session_dir = ray._private.utils.internal_kv_get_with_retry(
-                    self.get_gcs_client(),
-                    "session_dir",
-                    ray_constants.KV_NAMESPACE_SESSION,
-                    num_retries=ray_constants.NUM_REDIS_GET_RETRIES,
+            if (
+                node_to_connect_info is not None
+                and node_to_connect_info.temp_dir != self.temp_dir
+            ):
+                raise ValueError(
```

Comment:
can this happen?

Reply:
Good catch. The filter in `get_node_to_connect_to_for_driver` should catch this already.